Add missing Person and Number to verbs in the ewt-test

lapalme opened the pull request:
Use a script to add missing Person and Number to verbs, then patch 22 "unknown" manually.
For the moment, only the test corpus has been processed. If all goes well then I will do the same for the dev and train.
Please confirm that the process is acceptable and that it is with the right version.

PS: thanks for "resetting" my GitHub.

nschneid commented:
I think this setup will work. Thanks

lapalme commented:
Added missing Person and Number for the indicative verbs in the test/dev/train corpora

Most of these (about 10 000) were inserted using a Python script that found the Person and Number of the subject and inserted it for the verb. 

Some (a few hundred) that could not be identified were added manually.

In a few cases, this revealed bad head links or bad tenses that were corrected.

nschneid commented:
If you are done with the changes I can look over and then merge this pull request.

lapalme commented:
I think I am done with this set of this type of changes. If you find problems, do not hesitate to tell me.

PS: sorry for the previous PR on the master branch (I realized too late), you can delete it. 
I am still learning...

nschneid commented:
No worries! Git does have a steep learning curve.

Unfortunately I think the changes you've made are not based on the most recent version of the dev branch. Some changes committed earlier that are getting reverted—`NumType=Frac` on "half", `Typo=Yes` on "squirelled", and an instance of "it's" that is a typo of "its":

```diff
-14     half    half    NOUN    NN      Number=Sing|NumType=Frac        11      conj    11:conj:and     SpaceAfter=No
+14     half    half    NOUN    NN      Number=Sing     11      nummod  11:nummod       SpaceAfter=No
```

```diff
-5      are     be      AUX     VBP     Mood=Ind|Tense=Pres|VerbForm=Fin        6       aux:pass        6:aux:pass      _
-6      squirelled      squirrel        VERB    VBN     Tense=Past|Typo=Yes|VerbForm=Part|Voice=Pass    1       csubj   1:csubj CorrectForm=squirreled
+5      are     be      AUX     VBP     Mood=Ind|Number=Plur|Person=3|Tense=Pres|VerbForm=Fin   6       aux:pass        6:aux:pass      _
+6      squirelled      squirell        VERB    VBN     Tense=Past|VerbForm=Part|Voice=Pass     1       csubj   1:csubj _
```

```diff
-7      it's    its     PRON    PRP$    Gender=Neut|Number=Sing|Person=3|Poss=Yes|PronType=Prs|Typo=Yes 8       nmod:poss
       8:nmod:poss     CorrectForm=its
+7      it's    it's    PRON    PRP$    _       8       nmod:poss       8:nmod:poss     _
```

nschneid commented:
I'm not enough of a Git wizard to be sure, but what may have happened was that when I reset the lapalme:dev branch on the server, you still had the old files locally, and did a merge rather than discarding the old files.

lapalme commented:
Ouch..., I imagine that all the modifications described at
        https://github.com/UniversalDependencies/UD_English-EWT/commits/dev
(i.e. after April 30th) were not taken into account in "my" version.

What do you suggest ? Is there a way of "reappying" them  or making sure that I start with the last committed state

nschneid commented:
Maybe write a script to compare your version and the official dev version to see where there are differences other than the added person/number features? Or else reset the branch again, clone it to a clean directory, and re-add the features.

nschneid commented:
I guess the other way to do it would be a manual merge through git. So reset to the old dev branch your changes were originally based on, clone to a clean directory, copy your modified files into that directory, and then manually resolve merge conflicts from the more recent official dev branch. But I'm not sure if that would be easier.

nschneid commented:
Hi @lapalme, what is the status on this? It looks like just the dev set is updated so far.

lapalme commented:
Given my experience with my previous pull requests, I was waiting for a confirmation that now I was indeed working with the right version of "ewt-dev" corpus.
If you confirm that this is the case, I can apply the same kind of modifications to the test and train set of ewt.
thanks

nschneid commented:
The version you are modifying looks correct—I don't see others' changes being overridden. Thanks!

nschneid commented on en_ewt-ud-test.conllu:
```diff
@@ -28467,7 +28467,7 @@
 8	plumbers	plumber	NOUN	NNS	Number=Plur	5	obj	5:obj|10:nsubj	_
 9	who	who	PRON	WP	PronType=Rel	10	nsubj	8:ref	_
-10	visted	visit	VERB	VBD	Mood=Ind|Tense=Past|Typo=Yes|VerbForm=Fin	8	acl:relcl	8:acl:relcl	CorrectForm=visited
+10	visted	visit	VERB	VBD	Mood=Ind|Number=Sing|Person=3|Tense=Past|Typo=Yes|VerbForm=Fin	8	acl:relcl	8:acl:relcl	CorrectForm=visited
```
An interesting case: subject is a relative pronoun. I suppose this should be `Number=Plur` because the antecedent is plural? In the present tense it would matter for agreement:

- the plumber who visits / the plumbers who visit

lapalme replied:
right, it should be plural, there are probably a few other cases like these.

What do you suggest for "The team who work there are helpfull," (line 31963), should "work" be  "plural" (probably because 'are' is plural, even though the subject is singular. Those are the kind of things that should be decided manually.

nschneid replied:
Yes I would say that is plural because "team" is being interpreted collectively.

lapalme replied:
Hello

I have reviewed the "dev" and "test" for sentence with a relative pronoun as subject. It seems that these modifications have been pushed OK (see screen shoot).
I wanted to send a Pull Request for these, but I cannot find the "button" to do it...
![screenshot](https://user-images.githubusercontent.com/6200743/125881744-056ca925-b776-45e4-a4be-c1f221816132.jpg)

Do you see these modifications ?
thanks
Guy

nschneid replied:
Yes the modifications look good. They are modifications to the original pull request since I haven't merged it yet.

nschneid commented:
@lapalme Once you update the training set I think we'll be good to go!

lapalme commented:
That explains why I could not do a Pull request.
As the training set is much larger it will take a few days before I check the result of the script.

lapalme commented:
I have now updated the training set with many corrections, do you see them ? 

After using the script for adding Person and Number, I checked for relatives and corrected those that should be plural. I also looked at verbs at the 3rd person indicative present that did not end with "s" and realized that many coordination had not been dealt correctly. I corrected these also.

But in 18 cases, I was not sure what to do because I did not understand the sentence (English not being my mother tongue) or it seemed that the dependencies were not right, there were also other strange cases. I marked these with "ToCheck" in the MISC field. Perhaps you could have a look at these and suggest what to do in those cases or update these yourself.

thanks

nschneid commented:
Excellent, I see your changes! It will be a few days before I have a chance to go through and fix things, but I quickly looked at your ToCheck examples. Here are some comments in case you want to make the fixes:

- "ye" is an archaic form of "you". So second person, could be singular or plural.
- "between now and then Dunn, the AG have the time and resources to...": I assume this is missing "will" before "have". So "Dunn, the AG" (attorney general?) is the subject.
- "O Lord" is a vocative construction, so `vocative`. "O" might historically be a case marker but attaching it as `discourse` to "Lord" seems reasonable.
- "Merrist Wood...do HND in equine management": I had to look this up. Apparently [Merrist Wood](https://en.wikipedia.org/wiki/Guildford_College#Merrist_Wood_College) is a college and HND is a type of degree. I think this is the British use of plural agreement for a collective noun.
- "Put on a pair of spurs and before she makes that trot step, it's in the feeling when you feel her __think__ trot, tick her with the spur and lift with your body to keep her moving." I'm not sure I understand the meaning here but I would parse "her" as the raised subject of "think". So 3sg agreement looks right.
- "how good your bond it": I assume "it" is a typo for "is".
- "ordered that the ammunition...__use__": This is a subjunctive construction—I guess the verb is nonfinite? Cf. The judge ordered that he __be__ sentenced to a year in prison.
- "make 2 people __get__ married": causative construction; also nonfinite? Cf. make 2 people be married.
- "The difference is what they need to live": "is" should be the root of the sentence so "need" doesn't have 2 subjects.
- "Sussex have four toes" etc.: I think the subject is interpreted as plural, i.e. *Sussex breed chickens* have four toes.
- "you will need to [get his box and bowls, and whatever blanket and toys are his, and smell like home], and [__put__ him in a room].": infinitive
- "I would really love to hear what the French or those who had dated the French to say something about this." Sentence doesn't parse for me...I think the speaker begins one construction and ends up shifting to another. If I had to edit it: "I would really love to hear what the French or those who had dated the French **have** to say ~something~ about this."

lapalme commented:
Thanks for the suggestions. I have changed the annotations and removed most "ToCheck" in train, but there are still 7 left, because I am not too sure of the appropriate relations to use and between which words in the case of "repetition" or "missing words".

Two other strange case for me:
Line 120975 : this looks like a bad parse, but I do not know how to fix it
Line 179779 : you suggested to have "is" as root, but this is not what is usually done for a copula.

I also added 1 ToCheck in both dev and test. What do you suggest ?
Thanks

nschneid commented:
> Line 179779 : you suggested to have "is" as root, but this is not what is usually done for a copula.

https://universaldependencies.org/u/overview/simple-syntax.html:

> Exception: If the predicative element in the equation is a clause, then the copula verb is treated as the head of the clause, with the following clause as a ccomp (to prevent that the head of the smaller clause gets two subjects).

Does this apply? Or should "what" be left as the main predicate? I can never remember the rules for free relative clauses.

nschneid commented:
> Line 120975 : this looks like a bad parse, but I do not know how to fix it

"than they did 25 years ago": "did" should be the root of the embedded clause, with "they" as its subject and "ago" as its advmod dependent

nschneid commented on en_ewt-ud-train.conllu:
```diff
@@ -94583,11 +94583,12 @@
 11	the	the	DET	DT	Definite=Def|PronType=Art	12	det	12:det	_
 12	AG	ag	NOUN	NN	Number=Sing	9	appos	9:appos	_
-13	have	have	VERB	VBP	VerbForm=Inf	3	ccomp	3:ccomp	ToCheck=missingWill
+12.1	will	will	AUX	MD	VerbForm=Fin	_	_	13:aux	_
```
I don't think we're allowed to add a missing word like this: https://universaldependencies.org/u/overview/typos.html#missing-word These special nodes are only used for elided predicates.

lapalme replied:
I had understood that the [missing words were treated like ellipsis](https://universaldependencies.org/u/overview/typos.html#missing-word) which are then coded using "enhanced dependencies". At least, this is what seemed to be the appropriate way, if not what do you suggest for missing words.

nschneid replied:
My reading of that is that the tree should be treated as if there were ellipsis, and the policy on ellipsis is that null nodes are added only if it is *predicate* ellipsis. I don't know if there is a way to insert missing function words. I've opened an issue: UniversalDependencies/docs#801

nschneid commented on en_ewt-ud-train.conllu:
```diff
@@ -161088,9 +161089,9 @@
-4	what	what	PRON	WP	PronType=Int	0	root	0:root	_
+2	difference	difference	NOUN	NN	Number=Sing	3	nsubj	3:nsubj	_
+3	is	be	AUX	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	0	root	0:root	_
+4	what	what	PRON	WP	PronType=Int	3	cop	3:cop	_
```
Not cop

nschneid replied:
@amir-zeldes What is the correct structure here?

nschneid commented on en_ewt-ud-train.conllu:
```diff
@@ -179780,9 +179782,9 @@
+2	it	it	PRON	PRP	Case=Nom|Gender=Neut|Number=Sing|Person=3|PronType=Prs	3	nsubj	3:nsubj	_
 3	's	be	AUX	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	10	cop	10:cop	_
-4	not	not	PART	RB	_	10	advmod	10:advmod	_
+4	not	not	PART	RB	_	3	advmod	3:advmod	_
```
The original was correct for this sentence—copulas attaching as `cop` should not have dependents.

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -15043,7 +15043,7 @@
 1	My	my	PRON	PRP$	Number=Sing|Person=1|Poss=Yes|PronType=Prs	2	nmod:poss	2:nmod:poss	_
 2	wife	wife	NOUN	NN	Number=Sing	3	nsubj	3:nsubj|8:nsubj	_
-3	know	know	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	0	root	0:root	_
+3	know	know	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|Typo=Yes|VerbForm=Fin	0	root	0:root	CorrectForm=Yes
```
CorrectForm=knows

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -18275,8 +18275,8 @@
-13	have	have	AUX	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	15	aux	15:aux	_
-14	nt	not	PART	RB	_	15	advmod	15:advmod	_
+13	have	have	AUX	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|Typo=Yes|VerbForm=Fin	15	aux	15:aux	CorrectForm=has
+14	nt	not	PART	RB	Typo=Yes	15	advmod	15:advmod	CorrectForm=not
```
Because it's spelled together with "have" I'd say `CorrectForm=n't`

lapalme replied:
OK

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -18479,7 +18479,7 @@
 # text = Anyone know of any HHa training in Delaware ?
 1	Anyone	anyone	PRON	NN	Number=Sing	2	nsubj	2:nsubj	_
-2	know	know	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	0	root	0:root	_
+2	know	know	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|Typo=Yes|VerbForm=Fin	0	root	0:root	CorrectForm=knows
```
Actually this is not an error. This construction is short for "*Does* anyone know of...?".

lapalme replied:
So features should perhaps be "VerbForm=Inf" ?

nschneid replied:
Yeah

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -18510,7 +18510,7 @@
 18	Delaware	Delaware	PROPN	NNP	Number=Sing	16	obl	16:obl:to	_
 19	anyone	anyone	PRON	NN	Number=Sing	20	nsubj	20:nsubj	_
-20	know	know	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	3	parataxis	3:parataxis	_
+20	know	know	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|Typo=Yes|VerbForm=Fin	3	parataxis	3:parataxis	CorrectForm=knows
```
same

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -20528,17 +20528,17 @@
 2-3	your	_	_	_	_	_	_	_	_
 2	you	you	PRON	PRP	Case=Nom|Person=2|PronType=Prs	6	nsubj	6:nsubj|8:nsubj	_
-3	r	be	AUX	VBP	Mood=Ind|Number=Sing|Person=2|Tense=Pres|VerbForm=Fin	6	cop	6:cop	_
+3	r	be	AUX	VBP	Mood=Ind|Number=Sing|Person=2|Tense=Pres|Typo=Yes|VerbForm=Fin	6	cop	6:cop	CorrectForm=are
```
I'd say `CorrectForm='re`

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -20528,17 +20528,17 @@
 11	then	then	ADV	RB	PronType=Dem	13	advmod	13:advmod	_
-12	u	u	PRON	PRP	_	13	nsubj	13:nsubj	_
-13	need	need	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	0	root	0:root	_
+12	u	you	PRON	PRP	Case=Nom|Person=2|PronType=Prs|Typo=Yes	13	nsubj	13:nsubj	CorrectForm=you
```
Instead of `Typo=Yes` I'd say `Abbr=Yes`

lapalme replied:
OK

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -22337,7 +22337,7 @@
 1-2	whose	_	_	_	_	_	_	_	_
 1	who	who	PRON	WP	PronType=Int	4	nsubj	4:nsubj	_
-2	se	be	AUX	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	4	aux	4:aux	_
+2	se	be	AUX	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|Typo=Yes|VerbForm=Fin	4	aux	4:aux	CorrectForm=is
```
`CorrectForm='s`

lapalme replied:
OK

nschneid commented on en_ewt-ud-dev.conllu:
```diff
@@ -24643,8 +24643,8 @@
 9	but	but	CCONJ	CC	_	11	cc	11:cc	_
-10	u	u	PRON	PRP	_	11	nsubj	11:nsubj|13:nsubj:xsubj|18:nsubj:xsubj	_
-11	need	need	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	7	conj	7:conj:but	_
+10	u	u	PRON	PRP	Person=2|Poss=Yes|PronType=Prs|Typo=Yes	11	nsubj	11:nsubj|13:nsubj:xsubj|18:nsubj:xsubj	CorrectForm=you
```
Abbr=Yes

nschneid commented on en_ewt-ud-test.conllu:
```diff
@@ -17298,7 +17298,7 @@
 24	this	this	DET	DT	Number=Sing|PronType=Dem	25	det	25:det	_
 25	country	country	NOUN	NN	Number=Sing	27	nsubj	27:nsubj	_
-26	has	have	AUX	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	27	aux	27:aux	_
+26	has	have	AUX	VBZ	Mood=Ind|Number=Plur|Person=3|Tense=Pres|VerbForm=Fin	27	aux	27:aux	_
```
singular (country)

lapalme replied:
OK

nschneid commented on en_ewt-ud-train.conllu:
```diff
@@ -151291,7 +151291,7 @@
 13	your	you	PRON	PRP$	Person=2|Poss=Yes|PronType=Prs	14	nmod:poss	14:nmod:poss	_
 14	bond	bond	NOUN	NN	Number=Sing	15	nsubj	12:nsubj:xsubj|15:nsubj	_
-15	it	it	VERB	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	6	conj	1:csubj|6:conj:and	SpaceAfter=No|ToCheck=Verb!
+15	it	it	VERB	VBZ	Mood=Ind|Number=Sing|Person=3|Tense=Pres|Typo=Yes|VerbForm=Fin	6	conj	1:csubj|6:conj:and	CorrectForm=is|SpaceAfter=No
```
lemma = be

lapalme replied:
ok

nschneid commented on en_ewt-ud-train.conllu:
```diff
@@ -160968,10 +160968,10 @@
 4	trying	try	VERB	VBG	VerbForm=Ger	2	advcl	2:advcl:like	_
 5	to	to	PART	TO	_	6	mark	6:mark	_
-6	Make	make	VERB	VB	VerbForm=Inf|Typo=Yes	4	xcomp	4:xcomp	CorrectForm=make
+6	Make	make	VERB	VB	Typo=Yes|VerbForm=Inf	4	xcomp	4:xcomp	CorrectForm=make
```
Not sure if you added this but we don't consider it a typo if only the capitalization is weird.

lapalme replied:
I did not know that, for me it is a kind of Typo.

nschneid replied:
If we were only dealing with well-edited genres, then we could mark nonstandard capitalization. But in web data there are people who use all-caps or all-lowercase or other conventions. So we'd rather focus our efforts on typos other than miscapitalization, of which there are plenty!

nschneid commented on en_ewt-ud-train.conllu:
```diff
@@ -158529,7 +158529,7 @@
 22	early	early	ADJ	JJ	Degree=Pos	23	amod	23:amod	_
 23	60s	60	NOUN	NNS	Number=Plur	17	obl	17:obl:in	_
-24	use	use	VERB	VBP	Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	6	ccomp	6:ccomp	ToCheck=Number
+24	use	use	VERB	VBP	Mood=Sub|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin	6	ccomp	6:ccomp	_
```
VerbForm=Inf? I guess the subjunctive is a special use of the infinitive form.

lapalme replied:
OK

nschneid commented:
Hi @lapalme, are you waiting for additional feedback on changes discussed above? I can deal with the inserted "will" based on discussion we're having in another thread.

lapalme commented:
Yes I was waiting for  new feedback, I just committed a few modifications that take your suggestions into account.
If you can deal with the "missing" word problem, I would be happy to let you do this, because it seems to involve some "advanced" type of modifications.
